scripts/lint_prowjobs.go

```diff
@@ -0,0 +1,279 @@
+		Cluster:                      "prow-postsubmits-cluster",
+		DefaultServiceAccountName:    "postsubmits-build-account",
+		AthensHelmServiceAccountName: "charts-build-account",
+		DocsServiceAccountName:       "docs-build-account",
+		MakeTarget:                   "release",
+	}
+
+	pullBaseSha := os.Getenv("PULL_BASE_SHA")
+	pullPullSha := os.Getenv("PULL_PULL_SHA")
+	cmdOut, err := exec.Command(`git`, "diff", "--name-only", pullBaseSha, pullPullSha).Output()
+	if err != nil {
+		fmt.Fprintln(os.Stderr, "There was an error running the git command: ", err)
+		os.Exit(1)
+	}
+	output := string(cmdOut)
+	diffFiles := strings.Fields(output)
+	for _, file := range diffFiles {
+		if strings.Contains(file, "presubmits") {
+			presubmitFiles = append(presubmitFiles, file)
+		}
+		if strings.Contains(file, "postsubmits") {
+			postsubmitFiles = append(postsubmitFiles, file)
+		}
+	}
+
+	for _, presubmitFile := range presubmitFiles {
+		presubmitChecks(presubmitFile, presubmitConstants, &presubmitType)
+	}
+
+	for _, postsubmitFile := range postsubmitFiles {
+		postsubmitChecks(postsubmitFile, postsubmitConstants, &postsubmitType)
+	}
+
+	fmt.Println()
+	for presubmitFile, errors := range presubmitErrors {
+		if len(errors) > 0 {
+			fmt.Println(presubmitFile)
+			for _, errorString := range errors {
+				presubmitErrorCount++
+				fmt.Printf("  %s\n", errorString)
+			}
+		}
+		fmt.Println()
+	}
+
+	for postsubmitFile, errors := range postsubmitErrors {
+		if len(errors) > 0 {
+			fmt.Println(postsubmitFile)
+			for _, errorString := range errors {
+				postsubmitErrorCount++
+				fmt.Printf("  %s\n", errorString)
+			}
+		}
+		fmt.Println()
+	}
+	
+	if presubmitErrorCount > 0 || postsubmitErrorCount > 0 {
+		fmt.Println("❌ Validations failed!")
+		os.Exit(1)
+	} else {
+		fmt.Println("✅ Validations passed!")
+	}
+}
+
+func presubmitChecks(presubmitFilePath string, presubmitConstants PresubmitConstants, presubmitIface interface{}) {
+	presubmitErrors = make(map[string][]string)
+	presubmitFilePathSplit := strings.Split(presubmitFilePath, "/")
+	presubmitOrgRepo := presubmitFilePathSplit[1] + "/" + presubmitFilePathSplit[2]
+	presubmitFileName := presubmitFilePathSplit[3]
+
+	fileContents, err := ioutil.ReadFile(presubmitFilePath)
+	if err != nil {
+		fmt.Errorf("Unable to read contents of %s: %v", presubmitFilePath, err)
+	}
+	fileContentsString := string(fileContents)
+
+	if err := yaml.Unmarshal(fileContents, presubmitIface); err != nil {
+		fmt.Errorf("error unmarshaling %s: %v", presubmitFilePath, err)
+	}
+	presubmitJobs := presubmitIface.(*Presubmits)
+	presubmitType := presubmitJobs.PresubmitJobs[presubmitOrgRepo][0]
+
+	alwaysRunLineNo, _ := findLineNumber(fileContentsString, "always_run:")
+	if presubmitType.AlwaysRun {
+		errorString := alwaysRunLineNo + "\tPlease set always_run to false."
+		presubmitErrors[presubmitFileName] = append(presubmitErrors[presubmitFileName], errorString)
+	}
```

Comment:
I'd abstract these checks in some way where you can loop over them and individually test them. The same applies to postsubmits

```go
// A type for PreSubmitChecks
type preSubmitCheck func(presubmit prowapi.Presubmit, fileContent string, rule PresubmitConstants) (pass bool, message string)

// An function that returns a check function
func AlwaysRunCheck(desired bool) preSubmitCheck {
    return func(presubmit prowapi.Presubmit, fileContent string, rule PresubmitConstants) (pass bool, lineNo int, message string) {
        if presubmit.AlwaysRun != desired {
            return false, findLineNumber(fileContents, "always_run:"),  fmt.Sprintf("Please set always_run to %t", desired)
        }
        return true, 0, ""
    }
}

type preSubmitData struct{
    presubmit   prowapi.Presubmit
    filecontent string
    filename    string
}

// later
func main() {
    //...
    presubmits := []preSubmitData{/*  file data */}
    //... 
    presubmitChecks := []preSubmitCheck{
         AlwaysRunCheck(false),
         SkipReportCheck(false),
        JobBucketCheck(),
        // one function for each check
    }
    for _, presubmit := range presubmits {
       for _, check := range presubmitChecks {
           passed, lineNo, message := check(presubmit.presubmit, presubmit.fileContent, presubmit.filename)
           if !passed {
               // handle failing check
           }
       }
    
    }

}

```